src/Documentation/resources/Best_Practices.md

```diff
@@ -0,0 +1,421 @@
+    size and type is base on the *application domain model*.
+
+    -   Example: Users, Orders, etc.
+
+External Changing Data
+
+-   Grain are able to re-read the current state data from storage by using
+    State.ReadStateAsyc()
+
+-   A timer can also be used to re-read data from storage periodically as well
+
+    -   The functional requirements could be based on a suitable “staleness” of
+        the information
+
+        -   Example: Content Cache Grain
+
+-   Adding and Removing Fields
+
+    -   The storage provider will determine the effects of adding and removing
+        additional fields from its persisted state.
+
+    -   Azure table does not support schemas and should automatically adjust to
+        the additional fields.
+
+Writing Custom Providers
+
+-   Storage providers are simple to write which is also a significant extension
+    element for Orleans
+
+    -   Tutorial: *need tutorial*
+
+-   The API GrainState API contract drives the storage API contract (Write,
+    Clear, ReadStateAsync())
+
+-   The storage behavior is typically configurable (Batch writing, Hard or Soft
+    Deletions, etc.) and defined by the storage provider
+
+## Cluster Management
+
+-   Orleans automatically manages clusters
+
+    -   Worker roles (silo) can fail and join at any time which is handled
+        automatically
+
+    -   A silo instance table exists for diagnostics
+
+    -   There are also configuration options of an aggressive or a more lenient
+        failure detection
+
+-   Failures can happen at any time and are a normal occurrence
+
+    -   Lost grains are automatically reactivated
+
+    -   Grains that are making in-process calls can fail or timeout
+
+    -   Orleans provides it best effort for grain delivery
+
+    -   All of the network messages can be lost and should be retried by the application
+        if it is crucial
+
+        -   Common practice is to retry from end-to-end from the client/front
+            end
+
+-   Currently there is not a graceful shutdown
+
+    -   A Azure upgrade or reboot is treated as a node failure
+
+## Deployment and Production Management
+
+Service Monitoring
+
+-   Orleans provided information
+
+    -   Windows performance counters
+
+    -   Compact Azure metrics table
+
+    -   Extremely detailed Azure statistics table
+
+    -   Trace specific logging events
+
+-   Create a custom performance counters
+
+Scaling out and in
+
+-   Monitor the Service-Level Agreement (SLA)
+
+-   Add or Remove instances
+
+-   Orleans automatically rebalances and takes advantage of the new hardware
+
+## Logging and Testing
+
+-   Logging, Tracing, and Monitoring
+
+    -   GrainBase.GetLogger() exposes Info(), Warn(), Error(), Verbose()
+
+    -   .NET tracing and Runtimes trace to a local file by default
+
+    -   Azure Diagnostics is also easily consumed
+
+    -   Logger.LogConsumer.Add() allows the developer to choose their own
```

Comment:
This is outdated since we switched to leveraging `Microsoft.Extensions.Logging` in 2.0.